Doc/library/sqlite3.rst

```diff
@@ -325,6 +325,7 @@ Module functions
        enabling various :ref:`sqlite3-uri-tricks`.
 
    :param autocommit:
+       If *autocommit* is ``False``, a new transaction is implicitly opened.
        See :attr:`Connection.autocommit` and
        :ref:`sqlite3-transaction-control-autocommit` for more information.
        *autocommit* currently defaults to
```

Comment:
After some thought, I see several concerns here as a non-expert reader, that go beyond just the added sentence and highlight bigger issues with the parameter description as a whole (thus the lengthy comment):

* Mostly proximately, this appears to be missing a critical detail—that a new transaction is opened _when the `Connection` is created_; right now, as a lay reader, it would be quite non-obvious what this is actually trying to say if not for me having read the issue.
* Stepping back a pace or two, it seems that the "See also" should go either as the first line, or the last; the current ordering implies that the two referenced sections will describe the subject of the first sentence (as opposed to the parameter/attribute's behavior) in more detail, which in fact the former isn't mentioned at all in either section.
* More broadly, since the only mention of what this parameter actually _does_ is that a new transaction is implicitly opened if `False` is passed, as a non-expert user I would be rather confused reading this, and could easily think this is a (the?) primary effect of this parameter as opposed to just a rather specific behavior detail, potentially gaining a quite misleading impression of the parameter (and attribute's) purpose and importance.
* Overall, while we want to keep this parameter description as concise as possible, it's missing at least a brief one-line summary of this parameter's purpose (to control transaction handling behavior) to give the reader at least a basic idea of what it is actually _for_, and whether it might be prudent to read more about it in the linked references.

So, here's a start at addressing these issues:

```suggestion
       Controls :pep:`249` transaction handling behaviour.
       See :attr:`Connection.autocommit` and
       :ref:`sqlite3-transaction-control-autocommit` for more information.
       If ``False``, a new transaction is implicitly opened
       when the :class:`Connection` is created.
       *autocommit* currently defaults to
```

I'm still really not sure about mentioning it here at all, though. From my non-expert perspective, it feels like belaboring the otherwise fairly minimal parameter reference with one specific detail among many (arguably one of the least important of those mentioned) regarding the transaction handling behavior it controls. Furthermore, I'm really not sure how useful it is to most readers here, siloed away from the section that explains all the rest of the transaction handling behavior this parameter controls. IMO, it feels more likely to cause "premature confusion" mentioning it here than be actually helpful to most readers. If others agree, I would suggest just:

```suggestion
       Controls :pep:`249` transaction handling behaviour.
       See :attr:`Connection.autocommit` and
       :ref:`sqlite3-transaction-control-autocommit` for more information.
       *autocommit* currently defaults to
```

Reply:
Thanks for joining this discussion.

About the [Explanation section](https://docs.python.org/3.12/library/sqlite3.html?highlight=sqlite3#explanation), I also think that it is best for the reader to explain the full picture (`connect()`, `Connection.commit()`, and `Connection.rollback()`) in that single place. That is why I put the information there in the first place and I am constantly referring back to [MAL’s specification](https://github.com/python/cpython/issues/83638#issuecomment-1093853924) which is complete and concise.

Now about the [Reference section](https://docs.python.org/3.12/library/sqlite3.html?highlight=sqlite3#reference), I think that it should also have the full information on the implicit transaction opening behaviour (of `connect()`, `Connection.commit()`, and `Connection.rollback()`) because a reference should be complete. So I prefer your first commit suggestion. But since you also added the sentence ‘Controls :pep:`249` transaction handling behaviour.’ to the *autocommit* parameter, you should add a similar sentence ‘Controls legacy transaction handling behaviour.’ to the *isolation_level* parameter.

Reply:
CAM:

> I'm still really not sure about mentioning it here at all, though. From my non-expert perspective, it feels like belaboring the otherwise fairly minimal parameter reference with one specific detail among many (arguably one of the least important of those mentioned) regarding the transaction handling behavior it controls.

> IMO, it feels more likely to cause "premature confusion" mentioning it here than be actually helpful to most readers.

+1

I'm sceptical to including this in the explanation; the explanation already mentions that a transaction will always be open if autocommit is false. In my opinion, this is enough; how else could sqlite3 ensure that a transaction is always open?

Regarding the reference: it is not desirable that the reference explain every implementation detail; we want the reference consice, to the point. We simply want to equip the reader to be able to use the sqlite3 API, and that is of importance; we need to be able to balance the docs with enough, but not too many, implementation details.

Reply:
> I'm sceptical to including this in the explanation; the explanation already mentions that a transaction will always be open if autocommit is false. In my opinion, this is enough; how else could sqlite3 ensure that a transaction is always open?

Then why mentioning that `Connection.commit()` and `Connection.rollback()` implicitly open a new transaction? How else could `sqlite3` ensure that a transaction is always open?

I prefer that we list explicitly the *three* implicit-transaction-opening functions, but if we instead choose that is an ‘implementation detail’ then let’s be consistent and list no function *at all*. IMO the current partial listing situation is not satisfactory.

Reply:
> Then why mentioning that `Connection.commit()` and `Connection.rollback()` implicitly open a new transaction? How else could `sqlite3` ensure that a transaction is always open?

IMO, mentioning `commit()` and `rollback()` makes sense, because that is what a developer will _work with_ in manual commit mode (or transactional mode, as MAL calls it). With autocommit there are two modes:

- transactional (aka manual commit): we tell the user that a transaction is always open so they need to commit and rollback (explicitly) using the DB API methods `commit()` and `execute()`
- non-transactional (aka autocommit): the user does not have to think about transactions; SQLite takes care of everything.

I think this is very well explained in the current docs.

> But if we instead choose that it is an ‘implementation detail’ then let’s be consistent and list no function _at all_.

I'm strongly -1 to that suggestion.

Reply:
> IMO, mentioning commit() and rollback() makes sense, because that is what a developer will work with in manual commit mode (or transactional mode, as MAL calls it).

Yes that makes sense, and the developer will also work with `connect()` in this mode (and in the other mode).

> > But if we instead choose that it is an ‘implementation detail’ then let’s be consistent and list no function _at all_.
> 
> I'm strongly -1 to that suggestion.

I am also −1 to not listing any implicit-transaction-opening functions at all. But I am −2 to an incomplete listing. I don’t view the implicit-transaction-opening behaviours of `Connection.commit()` and `Connection.rollback()` as more important than `connect()`’s. An alternative implementation of the transactional mode (manual commit mode) for ensuring that a transaction is always open might have been to implicitly open a transaction at the first `Connection.execute()` or `Connection.executemany()` call following a `connect()`, `Connection.commit()`, or `Connection.rollback()` call, so to me the current implicit-transaction-opening behaviours of `connect()`, `Connection.commit()`, and `Connection.rollback()` are on an equal footing.

About the location of that information, I prefer to see it in both the Reference and Explanation sections like in the current text, but at least the information has to be *somewhere*.

Reply:
At least to me as a documentarian and not a subject-matter-expert, I guess I'm not really sure of the concrete harm in including a brief mention in the existing explanation to be more explicit that a transaction is opened on `connect()`. If it was a task-oriented how-to, a learning-oriented tutorial, or an austere reference, I could understand an objection based on length or appropriateness, like

> IMO, mentioning commit() and rollback() makes sense, because that is what a developer will work with in manual commit mode (or transactional mode, as MAL calls it).

Unlike a how-to, an explanation is *understanding*-oriented, and serves study rather than work and the conceptual rather than the concrete (and also tends to be the most generous in terms of word economy). To me, explicitly mentioning that a transaction is opened on `connect` would seem to serve the reader's conceptual understanding as to the behavior and implementation of `autocommit = False` mode.

 If it is decided to be mentioned, here's how I might do it:

```diff
diff --git a/Doc/library/sqlite3.rst b/Doc/library/sqlite3.rst
index 1681fc49e9..8baa4d4cd4 100644
--- a/Doc/library/sqlite3.rst
+++ b/Doc/library/sqlite3.rst
@@ -2390,9 +2390,9 @@ which implies :pep:`249`-compliant transaction control.
 This means:

 * :mod:`!sqlite3` ensures that a transaction is always open,
-  so :meth:`Connection.commit` and :meth:`Connection.rollback`
-  will implicitly open a new transaction immediately after closing
-  the pending one.
+  so :func:`connect`, :meth:`Connection.commit` and :meth:`Connection.rollback`
+  will implicitly open a new transaction
+  (immediately after closing the pending one, for the latter two).
   :mod:`!sqlite3` uses ``BEGIN DEFERRED`` statements when opening transactions.
 * Transactions should be committed explicitly using :meth:`!commit`.
 * Transactions should be rolled back explicitly using :meth:`!rollback`.
 ```
 
All that said, perhaps the greater practical cost in terms of word economy is that of discussing this point, on which I have been perhaps the most profligate spender... :joy:

Reply:
I prefer CAM's proposal in https://github.com/python/cpython/pull/99825#discussion_r1035377109. Géry, could you please change your PR to include this, and only this, change?

Reply:
Done.